### ISSUE TITLE
Add support for storage class

### ISSUE THREAD

**Issue by venezia** [samsung-cnct/kraken-lib#799]
In v1.7+ of kubernetes, annotations in PVCs like `volume.alpha.kubernetes.io/storage-class: default` are no longer supported.

One now has to create storage classes.  A storage class can be considered the default storage class, but an additional admission controller plugin is needed in order for this to work.

## Suggestion:
- [x] Add StorageClass to k2/kraken configuration.  
- [x] Create an AWS implementation.  
- [ ] Add AWS to the default configuration generated by `--generate`
- [ ] Add *_the ability to set a storage class to the cluster default_* should be available.  Perhaps `isDefault` is a boolean that can be available across all storageclasses.
- [ ] Configuration should warn user if multiple storage classes are configured to be the default
- [x] Add `DefaultStorageClass` admission controller plugin to `api-server.yaml`
- [x] Add Documentation

## References
StorageClass AWS [specs](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#aws) for configuration options that should be supported.
[Note](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1) mentioning the requirement of the admission controller
[Admission Controller Plugin List](https://kubernetes.io/docs/admin/admission-controllers/#defaultstorageclass) indicating the name and functionality of the admission controller plugin needed

## Samples
This has worked for me:
### API Server change
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: kube-apiserver
  namespace: kube-system
spec:
  hostNetwork: true
  containers:
  - name: kube-apiserver
    image: gcr.io/google_containers/hyperkube:v1.7.4
    command:
    - /hyperkube
    - apiserver
    - --etcd-servers=https://etcd.mbv5.internal:2379
    - --etcd-servers-overrides=/events#https://etcd.mbv5.internal:2379
    - --bind-address=0.0.0.0
    - --etcd-cafile=/etc/etcd/ssl/client-ca.pem
    - --etcd-certfile=/etc/etcd/ssl/client.pem
    - --etcd-keyfile=/etc/etcd/ssl/client-key.pem
    - --allow-privileged=true
    - --service-cluster-ip-range=10.32.0.0/12
    - --secure-port=443
    - --admission-control=DefaultStorageClass,NamespaceLifecycle,LimitRanger,ServiceAccount,ResourceQuota
    - --tls-cert-file=/etc/kubernetes/ssl/apiserver.pem
    - --tls-private-key-file=/etc/kubernetes/ssl/apiserver-key.pem
    - --client-ca-file=/etc/kubernetes/ssl/ca.pem
    - --service-account-key-file=/etc/kubernetes/ssl/service-account-key.pem
    - --cloud-provider=aws
    - --logtostderr=true
    - --authorization-mode=RBAC
    - --runtime-config=rbac.authorization.k8s.io/v1beta1=true
    - --v=2
    ports:
    - containerPort: 443
      hostPort: 443
      name: https
    - containerPort: 8080
      hostPort: 8080
      name: local
    volumeMounts:
    - mountPath: /etc/kubernetes/ssl
      name: ssl-certs-kubernetes
      readOnly: true
    - mountPath: /etc/etcd/ssl
      name: ssl-certs-etcd
      readOnly: true
    - mountPath: /etc/ssl/certs
      name: ssl-certs-host
      readOnly: true
  volumes:
  - hostPath:
      path: /etc/kubernetes/ssl
    name: ssl-certs-kubernetes
  - hostPath:
      path: /etc/etcd/ssl
    name: ssl-certs-etcd
  - hostPath:
      path: /usr/share/ca-certificates
    name: ssl-certs-host
  tolerations:
  - key: node-role.kubernetes.io/master
    operator: Equal
    effect: NoSchedule
```
### Storage Class
```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: gp2
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: kubernetes.io/aws-ebs
parameters:
  type: gp2
```
# Sample PVC request
```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: dummy4-pvc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 8Gi
```
#### How to know if it is working correctly
`kubectl get pvc` will show you if this is working right
##### Working incorrectly:
```bash
NAME         STATUS    VOLUME    CAPACITY   ACCESSMODES   STORAGECLASS   AGE
dummy4-pvc   Pending                                      default        4s
```
##### Working correctly:
```bash
NAME          STATUS    VOLUME                                     CAPACITY   ACCESSMODES   STORAGECLASS   AGE
dummy4-pvc    Bound     pvc-5bd690ba-927c-11e7-9405-1277328ef8e4   8Gi        RWO           gp2            16m
```

**Comment by davidewatson** [comment]
Labels /p1 or /p2 with /k2. I am unsure if all of our charts work with k8s 1.7 without this.

**Comment by notjames** [comment]
Really sorry folks. I am going to have to un-assign this. We found a quick workaround for our use-case and haven't the time to work on this project for now. So, reassigning to pool. This was the quick workaround I used along with the fact that the consul helm chart provides the rest.

```bash
#!/bin/sh


storage_config () {
cat <<EOF
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: cspvclaim
  accesModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3.5gi
parameters:
  type: gp2

EOF
}


storage_config | kubectl apply -f -
```

**Comment by guineveresaenger** [comment]
Per discussion with the team earlier, we are for now adding a default storage class to any AWS kraken cluster. `DefaultStorageClass` admission controller plugin has been added to default api-server.yaml as well as 1.7's api-server.yaml.

Adding the option to configure storage classes in the configuration file will be part of a future discussion.
PR: #862

**Comment by guineveresaenger** [comment]
PR was merged, and unfinished points are part of future work. Moving this to Done.

**Comment by coffeepac** [comment]
We should create a new issue with the parts of this issue that were not completed.  

Its good stuff, but not for right now.

**Comment by guineveresaenger** [comment]
#875 . Asking @venezia to comment/improve.

**Comment by leahnp** [comment]
@guineveresaenger can we confirm this works with K8s v1.6 and 1.5?

**Comment by guineveresaenger** [comment]
ToReview:
- PR was reviewed and merged
- Check if follow-up ticket covers the rest of the issue: #875 and is in the correct place in triage/backlog
- PR's already been merged; might be worthwhile to check if a 1.6 cluster comes up successfully. 1.6 and 1.5 should act similarly.

**Comment by leahnp** [comment]
LGTM https://github.com/samsung-cnct/kraken-lib/pull/862
